generator/internal/codegen_utils.cc

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "generator/internal/codegen_utils.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_replace.h"
+#include "absl/strings/str_split.h"
+#include <cctype>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+
+namespace {
+char const kGeneratorFileSuffix[] = ".gcpcxx.pb";
+}  // namespace
+
+std::string GeneratedFileSuffix() { return {kGeneratorFileSuffix}; }
+
+std::string LocalInclude(absl::string_view header) {
+  return absl::StrCat("\"", header, "\"");
+}
+
+std::string SystemInclude(absl::string_view header) {
+  return absl::StrCat("<", header, ">");
+}
+
+std::string CamelCaseToSnakeCase(absl::string_view input) {
```

Comment:
I am not sure if you like this version better, I do:


```cc
std::string SnakeIt2(absl::string_view in) {
  std::string output;
  output.reserve(in.size());
  bool needs_underscore = false;
  for (auto c : in) {
    auto const uc = static_cast<unsigned char>(c);
    if (c == '_') {
      needs_underscore = false;
      output += c;
      continue;
    }
    if (std::islower(uc)) {
      needs_underscore = true;
      output += c;
      continue;
    }
    if (needs_underscore) {
      output += '_';
      needs_underscore = false;
    }
    output += static_cast<char>(std::tolower(uc));
  }
  return output;
}

TEST(Snake, Base) {
  EXPECT_EQ("foo_bar_baz", SnakeIt2("FooBarBaz"));
  EXPECT_EQ("foo_bar_baz", SnakeIt2("fooBarBaz"));
  EXPECT_EQ("foo_bar_ba", SnakeIt2("fooBarBa"));
  EXPECT_EQ("foo_bar_baaaaa", SnakeIt2("fooBarBAAAAA"));
  EXPECT_EQ("foo_bar_b", SnakeIt2("foo_BarB"));
}
```

Reply:
Tried it, it changed "v1" to "v_1". Added additional test cases and reworked existing algorithm when it failed on "foo_BarB".

Reply:
ftr, I think we'd want "v_1", wouldn't we? `Foo123` -> `foo_123`??

Reply:
Ack, I am just leaving this here... your call:

```cc
std::string SnakeIt2(absl::string_view in) {
  std::string output;
  output.reserve(in.size());
  bool needs_underscore = false;
  for (auto c : in) {
    auto const uc = static_cast<unsigned char>(c);
    if (c == '_' || !std::isupper(uc)) {
      needs_underscore = c != '_';
      output += c;
      continue;
    }
    if (needs_underscore) output += '_';
    needs_underscore = false;
    output += static_cast<char>(std::tolower(uc));
  }
  return output;
}

TEST(Snake, Base) {
  EXPECT_EQ("v1", SnakeIt2("v1"));
  EXPECT_EQ("foo123", SnakeIt2("foo123"));
  EXPECT_EQ("foo123bar", SnakeIt2("foo123bar"));
  EXPECT_EQ("foo123_bar", SnakeIt2("foo123Bar"));
  EXPECT_EQ("foo_bar_baz", SnakeIt2("FooBarBaz"));
  EXPECT_EQ("foo_bar_baz", SnakeIt2("fooBarBaz"));
  EXPECT_EQ("foo_bar_ba", SnakeIt2("fooBarBa"));
  EXPECT_EQ("foo_bar_baaaaa", SnakeIt2("fooBarBAAAAA"));
  EXPECT_EQ("foo_bar_b", SnakeIt2("foo_BarB"));
  EXPECT_EQ("foo_bar_b___c", SnakeIt2("foo_BarB___C"));
  EXPECT_EQ("foo_bar_b___cd", SnakeIt2("foo_BarB___Cd"));
  EXPECT_EQ("foo_bar_b___c_d_e_f", SnakeIt2("foo_BarB___C_d_E_f"));
}
```